src/idlistener.c

```diff
@@ -0,0 +1,142 @@
+/*
+ *  Copyright (c) 2024 Filippo Rossoni
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "idlistener.h"
+#include "log.h"
+#include "http.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+#define MAX_REQUEST_SIZE 4096
+
+static void print_url(struct vpn_config *cfg) {
+	char url[512];
+	snprintf(url, sizeof(url), "https://%s:%d/remote/saml/start?redirect=1",
+			cfg->gateway_host, cfg->gateway_port);
+
+	if (cfg->realm[0] != '\0') {
+		strncat(url, "&realm=", sizeof(url) - 1);
+		char *dt = url + strlen(url);
+		url_encode(dt, cfg->realm);
+	}
+	log_info("Authenticate at %s\n", url);
+}
+
+// Function to parse HTTP request and extract parameter "id"
+static char* parse_request(const char *request) {
+	char *id_param;
+	char *query_start = strchr(request, '?');
+	if (query_start != NULL) {
+		id_param = strstr(query_start, "id=");
+		if (id_param != NULL) {
+			id_param += 3; // Length of "id="
+			char *id_end = strchr(id_param, '&');
+			if (id_end == NULL) {
+				id_end = strchr(id_param, ' ');
+			}
+			if (id_end == NULL) {
+				id_end = strchr(id_param, '\r');
+			}
+			if (id_end == NULL) {
+				id_end = id_param + strlen(id_param); // End of string
+			}
+			*id_end = '\0'; // Null-terminate the string
+			return strdup(id_param);
+
+		}
+	}
+	return NULL;
+}
+
+void send_response(int sockfd, const char *message) {
+	char response[MAX_REQUEST_SIZE];
+	sprintf(response, "HTTP/1.1 200 OK\r\n"
```

Comment:
Even though the `message` is coming from internal, maybe not everyone in future will be aware that there is a size restriction. At least the `sprintf` should be restricted to write out of bounds. Better would be to do a size check.